### PR TITLE
Highlighting of invalid indentation

### PR DESCRIPTION
**Pull request by micpiotrowski**
Maybe not the simplest and clearest one but it is: implementation of invalid indentation highlighting based on Machete Annotator class.
Probably better solution would be use a flexer to do this but something went wrong wen I try using it and I have no idea how to assign custom tooltip error text when mouse is hovered on selected error (see the first commit in this PR)

### CODE REVIEW

File: frontendFile/src/main/java/com/virtuslab/gitmachete/frontend/file/highlighting/MacheteAnnotator.java [hunk 1]
```diff
@@ -37,4 +45,129 @@ public void annotate(PsiElement element, AnnotationHolder holder) {
           .range(branch).create();
     }
   }
+
+  private void processIndentationElement(PsiElement element, AnnotationHolder holder) {
+    PsiElement parent = element.getParent();
+    assert parent != null : "Element has not parent";
```
Comment (PawelLipski) [inline frontendFile/src/main/java/com/virtuslab/gitmachete/frontend/file/highlighting/MacheteAnnotator.java:51]:
```suggestion
    assert parent != null : "Element has no parent";
```

File: frontendFile/src/main/java/com/virtuslab/gitmachete/frontend/file/highlighting/MacheteAnnotator.java [hunk 2]
```diff
@@ -37,4 +45,129 @@ public void annotate(PsiElement element, AnnotationHolder holder) {
...
+      holder.newAnnotation(HighlightSeverity.ERROR,
+          "Indentation width is not multiple of ${indentationParameters.indentationWidth}" +
+              " as first indented line suggests")
+          .range(element).create();
+    }
+
+    thisLevel = thisIndentationText.length() / indentationParameters.indentationWidth;
+
+    if (hasPrevLevelCorrectWidth && thisLevel > prevLevel + 1) {
+      holder.newAnnotation(HighlightSeverity.ERROR, "Indentation level does not math")
```
Comment (PawelLipski) [inline frontendFile/src/main/java/com/virtuslab/gitmachete/frontend/file/highlighting/MacheteAnnotator.java:97]: `match`

File: frontendFile/src/main/grammar/Machete.bnf [hunk 3]
```diff
@@ -17,7 +17,7 @@ simpleFile ::= item_*
 
 // Even if we don't have comments in our machete file this element must be here - language parser definition
```
Comment (PawelLipski) [inline frontendFile/src/main/grammar/Machete.bnf:18]: Nit:
```suggestion
// Even if we don't have comments in our machete file, the COMMENT element must be here - language parser definition
```

File: frontendFile/src/main/java/com/virtuslab/gitmachete/frontend/file/highlighting/MacheteAnnotator.java [hunk 4]
```diff
@@ -37,4 +45,129 @@ public void annotate(PsiElement element, AnnotationHolder holder) {
...
+      hasPrevLevelCorrectWidth = true;
+    } else {
+      var prevIndentationText = prevIndentationNodeOption.get().getText();
+      hasPrevLevelCorrectWidth = prevIndentationText.length() % indentationParameters.indentationWidth == 0;
+      prevLevel = prevIndentationText.length() / indentationParameters.indentationWidth;
+    }
+
+    var thisIndentationText = element.getText();
+
+    if (thisIndentationText.chars().filter(c -> c != indentationParameters.indentationCharacter).count() > 0) {
```
Comment (PawelLipski) [inline frontendFile/src/main/java/com/virtuslab/gitmachete/frontend/file/highlighting/MacheteAnnotator.java:81]: Nit: `xd.filter(f).count > 0` can likely be replaced with `xd.exists(f)` or sth like that

File: frontendFile/src/main/java/com/virtuslab/gitmachete/frontend/file/highlighting/MacheteAnnotator.java [hunk 5]
```diff
@@ -37,4 +45,129 @@ public void annotate(PsiElement element, AnnotationHolder holder) {
...
+      holder.newAnnotation(HighlightSeverity.ERROR, "Indentation character does not match the first indented line")
```
Comment (PawelLipski) [inline frontendFile/src/main/java/com/virtuslab/gitmachete/frontend/file/highlighting/MacheteAnnotator.java:82]: Would be nice if the message also mentions what's the expected & what's the actual indent character

File: frontendFile/src/main/java/com/virtuslab/gitmachete/frontend/file/highlighting/MacheteAnnotator.java (hunk 2 above)
Comment (PawelLipski) [reply frontendFile/src/main/java/com/virtuslab/gitmachete/frontend/file/highlighting/MacheteAnnotator.java:97]: Maybe even for super-clarity sth like:
```suggestion
      holder.newAnnotation(HighlightSeverity.ERROR, "Too much indent on this line")
```

File: frontendFile/src/main/java/com/virtuslab/gitmachete/frontend/file/highlighting/MacheteAnnotator.java (hunk 4 above)
Comment (micpiotrowski) [reply frontendFile/src/main/java/com/virtuslab/gitmachete/frontend/file/highlighting/MacheteAnnotator.java:81]: `exists()` seems to not exists for `IntStream` :/

Comment (micpiotrowski) [reply frontendFile/src/main/java/com/virtuslab/gitmachete/frontend/file/highlighting/MacheteAnnotator.java:81]: Oh, ok, I can use `anyMatch` instead

Comment (micpiotrowski) [reply frontendFile/src/main/java/com/virtuslab/gitmachete/frontend/file/highlighting/MacheteAnnotator.java:81]: Anyway, I done this in another way

File: frontendFile/src/main/java/com/virtuslab/gitmachete/frontend/file/highlighting/MacheteAnnotator.java [hunk 6]
```diff
@@ -78,8 +78,11 @@ private void processIndentationElement(PsiElement element, AnnotationHolder hold
 
     var thisIndentationText = element.getText();
 
-    if (thisIndentationText.chars().filter(c -> c != indentationParameters.indentationCharacter).count() > 0) {
-      holder.newAnnotation(HighlightSeverity.ERROR, "Indentation character does not match the first indented line")
+    var wrongIndentChar = thisIndentationText.chars().filter(c -> c != indentationParameters.indentationCharacter).findFirst();
```
Comment (PawelLipski) [inline frontendFile/src/main/java/com/virtuslab/gitmachete/frontend/file/highlighting/MacheteAnnotator.java:81]: Hmm actually what happens when the file is empty or there's no indent on any line?

Comment (micpiotrowski) [reply frontendFile/src/main/java/com/virtuslab/gitmachete/frontend/file/highlighting/MacheteAnnotator.java:81]: Nothing special - this piece of code will even not being executed

**Pull request merged**